lib/foreman/renderer.rb

```diff
@@ -1,389 +1,48 @@
-    end
-
-    def host_param_false?(name)
-      check_host
-      @host.params.has_key?(name) && Foreman::Cast.to_bool(@host.params[name]) == false
-    end
-
-    def subnet_has_param?(subnet, param_name)
-      validate_subnet(subnet)
-      subnet.parameters.exists?(name: param_name)
-    end
-
-    def subnet_param(subnet, param_name)
-      validate_subnet(subnet)
-      param = subnet.parameters.where(name: param_name).first
-      param.nil? ? nil : param.value
-    end
+    class << self
+      attr_writer :config
 
-    def render_safe(template, allowed_methods = [], allowed_vars = {}, scope_variables = {})
-      if Setting[:safemode_render]
-        box = Safemode::Box.new self, allowed_methods, template_name
-        erb = ERB.new(template, nil, '-')
-        box.eval(erb.src, allowed_vars.merge(scope_variables))
-      else
-        # we need to keep scope variables and reset them after rendering otherwise they would remain
-        # after snippets are rendered in parent template scope
-        kept_variables = {}
-        scope_variables.each { |k, v| kept_variables[k] = instance_variable_get("@#{k}") }
-
-        allowed_vars.merge(scope_variables).each { |k, v| instance_variable_set "@#{k}", v }
-        erb = ERB.new(template, nil, '-')
-        # erb allows to set location since Ruby 2.2
-        erb.location = template_name, 0 if erb.respond_to?(:location=)
-        result = erb.result(binding)
-
-        scope_variables.each { |k, v| instance_variable_set "@#{k}", kept_variables[k] }
-        result
+      def config
+        @config ||= Foreman::Renderer::Configuration.new
       end
-    rescue ::Racc::ParseError, ::SyntaxError => e
-      # Racc::ParseError is raised in safe mode, SyntaxError in unsafe mode
-      new_e = Foreman::Renderer::SyntaxError.new(N_("Syntax error occurred while parsing the template %{template_name}, make sure you have all ERB tags properly closed and the Ruby syntax is valid. The Ruby error: %{message}"), :template_name => template_name, :message => e.message)
-      new_e.set_backtrace(e.backtrace)
-      raise new_e
-    end
-
-    def foreman_server_fqdn
-      config = URI.parse(Setting[:foreman_url])
-      config.host
-    end
-
-    def foreman_server_url
-      Setting[:foreman_url]
-    end
 
-    class_eval do
-      [:debug, :info, :warn, :error, :fatal].each do |level|
-        define_method("log_#{level}".to_sym) do |msg|
-          template_logger.send(level, msg) if msg.present?
-        end
+      def configure
+        yield config
       end
-    end
 
-    def template_name
-      @template_name || 'Unnamed'
-    end
-
-    def pxe_kernel_options
-      return '' unless @host || @host.operatingsystem
-      @host.operatingsystem.pxe_kernel_options(@host.params).join(' ')
-    rescue => e
-      template_logger.warn "Unable to build PXE kernel options: #{e}"
-      ''
-    end
+      def render_template(template: nil, subjects: {}, params: {}, variables: {})
```

Comment:
@ares thanks for the feedback. We actually want to create source instance so in our Git Template plugin we can override this method and modify creation of the source according to our need. Otherwise we would have to create source and scope in multiple places (`Host::Base#render_template`,  `Host::Managed#diskLayout`, etc.) which goes against our initial though about having single helper method for rendering). Also If we pass source and scope from outside (like you suggested) this method mostly lose its point (as it would just pass data down the chain and nothing else).

If you don't like the subject scope we can remove it and replace with 3 keyword arguments:

- host
- scope
- source

Please let us know what you think about it. @timogoebel it would also be great to hear your thoughts :).

Reply:
after Friday discussion and weekend for thinking about this more and morning quick sync with @timogoebel I think we don't have to change this too much,

I'd like to ask for extracting scope/sources out of subjects. Bonus points for extracting instantiation out of this method like @kamils-iRonin suggests in gist like this

```ruby
def render_template(template: nil, subjects: {}, params: {}, variables: {})
  source = Foreman::Renderer.create_source(host, template) # here, one could check host's params to know what source type should be created
  scope = Foreman::Renderer.create_scope(host: subjects[:host], params: params, variables: variables)
  Foreman::Renderer.render_template(source, scope)
end
```

That would still give you power of overriding easily but you wouldn't have to override the renderer, just this method. And in future, we can extend the core so we'd formalize how to map source to host parameters.